### PR TITLE
Split k mm fix

### PR DESCRIPTION
**Pull request by xiaohuguo2023**
@scxiao @wenchenvincent @jayfurmanek @vgokhale 

While @scxiao mentioned we need adapt our autune system to enable SPLIT_K working properly, Jason suggest I should make a pull request for all to review where we should start with. 

The current changes have been summarized below:

./script/amd/gemm/matmul.py

 Fixing SPLIT_K matmul

    - add pytest coverage when using SPLIT_K.

    - add maxloc function to print out the max value discrepancy

    -  fix output matrix initialization issue

    - change the leaky_relu function to be same with torch.nn.functional.leaky_relu, and add pytest coverage when using activation leaky_relu

./python/tutorials/03-matrix-multiplication.py
- change leaky_relu definition to be same with  torch.nn.functional.leaky_relu and add test coverage if doing fused matmul

### CODE REVIEW

File: scripts/amd/gemm/matmul.py [hunk 1]
```diff
@@ -185,25 +187,24 @@ def matmul_kernel_splitK(
     # while the accumulator is still in FP32!
     if ACTIVATION == "leaky_relu":
         accumulator = leaky_relu(accumulator)
-    c = accumulator.to(tl.float16)
```
Comment (vgokhale) [inline scripts/amd/gemm/matmul.py:188]: I'm curious why we don't want to convert to f16? Is it for the atomic_add?

Comment (vgokhale) [reply scripts/amd/gemm/matmul.py:188]: So is this just for splitK > 1? If yes, can we do this conditionally if we use splitk or not?

File: python/tutorials/03-matrix-multiplication.py [hunk 2]
```diff
@@ -336,7 +335,9 @@ def test_correctness(M, N, K, in_dtype, out_dtype):
     a = torch.randn((M, K), device='cuda', dtype=torch.float16)
     b = torch.randn((K, N), device='cuda', dtype=torch.float16)
     triton_output = matmul(a, b)
+    triton_relu_output = matmul(a, b, activation="leaky_relu")
     torch_output = torch.matmul(a, b)
+    torch_relu_output = torch.nn.functional.leaky_relu(torch_output)
```
Comment (scxiao) [inline python/tutorials/03-matrix-multiplication.py:340]: What is the purpose of the change here?

Comment (xiaohuguo2023) [reply python/tutorials/03-matrix-multiplication.py:340]: yeah, @scxiao ,you are right, I don't need temp variable for later compare.

Comment (scxiao) [reply python/tutorials/03-matrix-multiplication.py:340]: My question is: what is the reason of adding the activation function here?

Comment (xiaohuguo2023) [reply python/tutorials/03-matrix-multiplication.py:340]: This case is for fused matmul. it has activation but not tested in the pytest

File: scripts/amd/gemm/matmul.py [hunk 3]
```diff
@@ -218,7 +219,7 @@ def matmul(a, b, activation=""):
     M, K = a.shape
     K, N = b.shape
     # Allocates output.
-    c = torch.empty((M, N), device=a.device, dtype=a.dtype)
+    c = torch.zeros((M, N), device=a.device, dtype=torch.float32)
```
Comment (scxiao) [inline scripts/amd/gemm/matmul.py:222]: We cannot always use fp32 since this will affect performance. so this line should be change to:
`c = torch.zeros((M, N), device=a.device, dtype=a.dtype)`

File: scripts/amd/gemm/matmul.py [hunk 4]
```diff
@@ -185,25 +187,24 @@ def matmul_kernel_splitK(
     # while the accumulator is still in FP32!
     if ACTIVATION == "leaky_relu":
         accumulator = leaky_relu(accumulator)
-    c = accumulator.to(tl.float16)
+    c = accumulator
```
Comment (scxiao) [inline scripts/amd/gemm/matmul.py:190]: This change should revert back.

Comment (xiaohuguo2023) [reply scripts/amd/gemm/matmul.py]: If I revert this line back, 16,16, 1026 is ok, but not for 32, 32, 1026, see below

✅ Triton and Torch match for size, (M: 16, N: 16, K: 1026)
Max difference location (row, column): tensor([ 0, 15])
Value at maxloc in triton output: -73.25
Value at maxloc in torch ouput: -73.3125

❌ Triton and Torch differ for size, (M: 32, N: 32, K: 1026)
Max difference location (row, column): tensor([ 3, 23])
Value at maxloc in triton output: -71.8125
Value at maxloc in torch ouput: -71.875

File: scripts/amd/gemm/matmul.py (hunk 1 above)
Comment (xiaohuguo2023) [reply scripts/amd/gemm/matmul.py:188]: I think I know understand why I don't want to convert to f16 now. 

If I revert this line back, 16,16, 1026 is ok, but not for 32, 32, 1026, see below

✅ Triton and Torch match for size, (M: 16, N: 16, K: 1026)
Max difference location (row, column): tensor([ 0, 15])
Value at maxloc in triton output: -73.25
Value at maxloc in torch ouput: -73.3125

❌ Triton and Torch differ for size, (M: 32, N: 32, K: 1026)
Max difference location (row, column): tensor([ 3, 23])
Value at maxloc in triton output: -71.8125
Value at maxloc in torch ouput: -71.875

File: scripts/amd/gemm/matmul.py (hunk 3 above)
Comment (xiaohuguo2023) [reply scripts/amd/gemm/matmul.py]: done!

**Comment by zhanglx13** [comment]
@xiaohuguo2023  What is the status of this PR. Do we have any conclusions?

**Comment by xiaohuguo2023** [comment]
@zhanglx13 can we wait for fp16 atomic-add ISA instruction fix, we may need redo benchmarking for splitK

cc. @wenchenvincent

**Comment by zhanglx13** [comment]
Do we have a timeline for the isa fix?

**Comment by xiaohuguo2023** [comment]
> Do we have a timeline for the isa fix?

we have a ticket for this,  working on it. hopefully by next week.

 https://github.com/ROCm/frameworks-internal/issues/7465

**Comment by vgokhale** [comment]
@xiaohuguo2023 can we close this PR? This work will be superseded by your next task on splitk.
